### PR TITLE
Integrated Mapster and added tests

### PR DESCRIPTION
**Pull request by Arenooo**
Integrated Mapster and added tests
---
- Added mapster
- Added tests
- Added DTO-s and requests
- Fixed data services

### CODE REVIEW

File: WA.Pizza.API/Program.cs [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
+using WA.Pizza.Infrastructure.Data;
+using WA.Pizza.Infrastructure.Data.Services;
+
 var builder = WebApplication.CreateBuilder(args);
 
 // Add services to the container.
 
-builder.Services.AddControllers();
+//builder.Services.AddControllers();
+builder.Services.AddControllersWithViews();
```
Comment (dcwaas) [inline WA.Pizza.API/Program.cs:11]: WithViews required for MVC, we creating API project

File: WA.Pizza.Infrastructure/Data/Services/BasketDataService.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace WA.Pizza.Infrastructure.Data.Services
+{
+    public class BasketDataService
+    {
+        private PizzaContext _dbContext;
+
+        public BasketDataService(PizzaContext dbContext)
+        {
+            MappingConfigurer.Configure();
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/Data/Services/BasketDataService.cs:16]: Move it to program to do it once and for all services

File: WA.Pizza.Infrastructure/MappingConfigurer.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            .Map(d => d.Quantity, s => s.Quantity);
+        /*TypeAdapterConfig<Basket, BasketDTO>.NewConfig()
+            .Map(d => d.Id, s => s.Id)
+            .Map(d => d.Items, s => s.Items);
+
+        TypeAdapterConfig<BasketDTO, Basket>.NewConfig()
+            .Map(d => d.Items, s => s.Items);
+
+        TypeAdapterConfig<Order, OrderDTO>.NewConfig()
+            .Map(d => d.Id, s => s.Id)
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/MappingConfigurer.cs:117]: ?? What it is here for?

File: WA.Pizza.Core/Exceptions/WAPizzaFailedToCreateException.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace WA.Pizza.Core.Exceptions
+{
+    public class WAPizzaFailedToCreateException : WAPizzaException
```
Comment (dcwaas) [inline WA.Pizza.Core/Exceptions/WAPizzaFailedToCreateException.cs:9]: Isn't it too much customexceptions types, when you can pass details in message?

File: WA.Pizza.Core/Models/Items/CatalogItem.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Threading.Tasks;
+
+namespace WA.Pizza.Core.Models.Items
+{
+    public class CatalogItem
+    {
+        public int Id { get; set; }
+        public string Description { get; set; }
+        public string Name { get; set; }
+        public decimal Price { get; set; }
```
Comment (dcwaas) [inline WA.Pizza.Core/Models/Items/CatalogItem.cs:14]: You already created ItemBase, why this class doesn't have it as parent?

File: WA.Pizza.Core/Models/Order.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        public decimal Price { get; set; }
+        public DateTime Date { get; set; }
+        public OrderStatus Status { get; set; }
+
+        public enum OrderStatus
+        {
+            InProgress,
+            Delivered,
+            Cancelled
+        }
```
Comment (dcwaas) [inline WA.Pizza.Core/Models/Order.cs:23]: Move it to separate file

File: WA.Pizza.Infrastructure/DTO/Basket/AddBasketItemRequest.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿namespace WA.Pizza.Infrastructure.DTO.Basket;
+
+public record AddBasketItemRequest()
+{
+    public int Id { get; set; }
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/DTO/Basket/AddBasketItemRequest.cs:5]: Discussed on meeting about Id field in requests

File: WA.Pizza.Infrastructure/Data/PizzaContext.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        public DbSet<Basket> Baskets { get; set; }
+        public DbSet<Order> Orders { get; set; }
+        public DbSet<BasketItem> BasketItems { get; set; }
+        public DbSet<OrderItem> OrderItems { get; set; }
+
+        public PizzaContext(DbContextOptions<PizzaContext> options) : base(options) { }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            var catalogItem = modelBuilder.Entity<CatalogItem>();
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/Data/PizzaContext.cs:25]: No needs to declare and store for each entity, example:
```
modelBuilder.Entity<CatalogItem>(b => 
{
     b.HasKey(m => m.Id);
     ....
}
```

File: WA.Pizza.Infrastructure/Data/Services/OrderDataService.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+            var order = new Order
+            {
+                Date = DateTime.Now,
+                Price = basket.Items.Sum(m => m.Price * m.Quantity),
+                Items = basket.Items.Adapt<List<OrderItem>>(),
+                Status = Order.OrderStatus.InProgress,
+            };
+            
+            var o = _dbContext.Orders.Add(order);
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/Data/Services/OrderDataService.cs:57]: t n i c
(this naming is confusing)
You can return order.Id after saving

File: WA.Pizza.Infrastructure/Data/Services/BasketDataService.cs (hunk 2 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/Data/Services/BasketDataService.cs:16]: +

File: WA.Pizza.Infrastructure/MappingConfigurer.cs (hunk 3 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/MappingConfigurer.cs:117]: \+

File: WA.Pizza.Core/Models/Items/CatalogItem.cs (hunk 5 above)
Comment (Arenooo) [reply WA.Pizza.Core/Models/Items/CatalogItem.cs:14]: \+

File: WA.Pizza.Core/Exceptions/WAPizzaFailedToCreateException.cs (hunk 4 above)
Comment (Arenooo) [reply WA.Pizza.Core/Exceptions/WAPizzaFailedToCreateException.cs:9]: Should I leave only the base exception?

File: WA.Pizza.Core/Models/Order.cs (hunk 6 above)
Comment (Arenooo) [reply WA.Pizza.Core/Models/Order.cs:23]: +

File: WA.Pizza.Infrastructure/DTO/Basket/AddBasketItemRequest.cs (hunk 7 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/DTO/Basket/AddBasketItemRequest.cs:5]: +

File: WA.Pizza.Infrastructure/Data/PizzaContext.cs (hunk 8 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/Data/PizzaContext.cs:25]: +

File: WA.Pizza.Infrastructure/Data/Services/OrderDataService.cs (hunk 9 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/Data/Services/OrderDataService.cs:57]: \+

File: WA.Pizza.API/Program.cs (hunk 1 above)
Comment (Arenooo) [reply WA.Pizza.API/Program.cs]: \+

**Review by dcwaas — APPROVED** [review]

**Pull request merged**